### ISSUE TITLE
How to pass string between Host and WASM using c-api

### ISSUE THREAD

**Issue by mrpre** [wasmerio/wasmer#526]
Need:
1 Pass string(char *) from host to `wasm`.  
2 Pass string(char *) from `wasm` to host.  

For example   

Host code like this:
```
...
wasm_func = instance.export[0];
char *host_param_string = "host string";
char *wasm_return_string = call_wasm_function(instance, wasm_func, host_param_string);
...
```

wasm code like this:
```
extern host_func();
char *wasm_param_string = "wasm string";
char *host_return_string = call_host_function(host_func, wasm_param_string);
```

That means I need four-direction of interactions  
(1). Pass string from host to `wasm` and get string in `wasm`.   
(2). Get the return string from `wasm` in host.  
(3). Pass string from `wasm` to host and get string in host.  
(4). Get the return string from host in `wasm`.    

Seems the example`lib/runtime-c-api/tests/test-import-function.c` only matches (3) and `lib/runtime-c-api/tests/test-exported-memory.c` only matches (2).
Any suggestion?

**Comment by MarkMcCaskey** [comment]
Hello!  This is a good question!

As host memory is not available to the Wasm module, the only way to make 1 and 4 work is to allocate the memory on the Wasm side and have the host write to it.  A good way to do this is to export a `malloc` function from your Wasm module and have the host call that as-needed.  This is a very general solution.

If you're using this for a more specialized use case and you're writing your own version of malloc on the Wasm side anyways, you can always save certain chunks of memory as memory that the host can use as it wants or have predefined locations for storing things.

There are many ways to do this actually, a hypothetical one, which I don't recommend in practice, is to use an iterator-style API where the guest can ask for "one more character". And then fill its own buffer one host-call at a time.  This is pretty simple to implement but has a lot of overhead.  You could group these together and say "give me n more bytes" and pass a pointer with enough space for the host to write it.  But stateful APIs can be error-prone.

Anyways, I think you get the general idea.  Making things like this easy is something we're thinking about.  The short-term solution is documentation (adding 1 & 4 to the examples is a good idea!) but in the future I think it probably makes sense to have some sensible defaults/easy ways to get started here.

The tl;dr is:
- you need to allocate on the Wasm side
- you have to copy things from host memory to get it into guest memory (how you do so is up to you)

**Comment by mrpre** [comment]
Quite clear answer !

**Comment by mrpre** [comment]
How about the `wasmer/lib/emscripten/`
